### PR TITLE
Prevent loss to be moved to the cpu before backward call.

### LINKED ISSUE #9296: Loss not on correct device when training with move_metrics_to_cpu=True
## 🐛 Bug
`move_metrics_to_cpu=True` seems to move loss to CPU which results in an error when training with native mixed precision.
This is related to the original issue reported in https://github.com/MIC-DKFZ/nnDetection/issues/25 and did not occur with other lightning versions (older version and/or move_metrics_to_cpu=False work fine).

Error when training with mixed precision and `move_metrics_to_cpu=True`:
```
/usr/local/lib/python3.7/dist-packages/torch/cuda/amp/grad_scaler.py in scale(self, outputs)
    159         # Short-circuit for the common case.
    160         if isinstance(outputs, torch.Tensor):
--> 161             assert outputs.is_cuda or outputs.device.type == 'xla'
    162             if self._scale is None:
    163                 self._lazy_init_scale_growth_tracker(outputs.device)
```

### To Reproduce

Can be reproduced with the boring model in colab by passing the following flags to the trainer:
```
        precision=16,  # native mixed precision
        move_metrics_to_cpu=True,
        gpus=[0],  # use GPU
```

### Expected behavior
No error :) 

### Environment

- PyTorch Lightning Version (e.g., 1.3.0):
- PyTorch Version (e.g., 1.8)
- Python version:
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #9296

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 1]
```diff
@@ -616,3 +617,17 @@ def validation_step(self, batch, batch_idx):
     )
 
     trainer.fit(model)
+
+
+@RunIf(min_gpus=1)
+def test_move_metrics_to_cpu(tmpdir):
```
Comment (carmocca) [inline tests/trainer/logging_/test_eval_loop_logging.py:623]: Why did you put this in `eval_loop_logging` if it's training related and not logging related?

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 2]
```diff
@@ -616,3 +617,17 @@ def validation_step(self, batch, batch_idx):
...
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        fast_dev_run=True,
+        amp_backend="native",
+        precision=16,
+        move_metrics_to_cpu=True,
+        gpus=1,
+    )
+    trainer.fit(BoringModel())
```
Comment (carmocca) [inline tests/trainer/logging_/test_eval_loop_logging.py:633]: can you assert something? perhaps the result.minimize device in `on_before_backward`

**Review by carmocca — APPROVED** [review]

File: tests/trainer/logging_/test_eval_loop_logging.py (hunk 1 above)
Comment (tchaton) [reply tests/trainer/logging_/test_eval_loop_logging.py:623]: moved to train.

File: tests/trainer/logging_/test_eval_loop_logging.py (hunk 2 above)
Comment (tchaton) [reply tests/trainer/logging_/test_eval_loop_logging.py]: Done !

**Review by kaushikb11 — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
